include/rcpputils/time.hpp

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCPPUTILS__TIME_HPP_
+#define RCPPUTILS__TIME_HPP_
+
+#include <chrono>
+
+#include "rcutils/time.h"
+
+namespace rcpputils
+{
+
+/// Convert to std::chrono::nanoseconds.
+/**
+ * This function help to convert from std::chrono::duration to std::chrono::nanoseconds and throw
+ * exception if overflow occurs while coverting.
+ *
+ * \param[in] time The time to be converted to std::chrono::nanoseconds.
+ * \return std::chrono::nanoseconds.
+ * \throws std::invalid_argument if time is bigger than std::chrono::nanoseconds::max() or less than
+ * std::chrono::nanoseconds::min().
+ */
+template<typename DurationRepT = int64_t, typename DurationT = std::milli>
+std::chrono::nanoseconds convert_to_nanoseconds(
+  const std::chrono::duration<DurationRepT, DurationT> & time)
+{
+  // Casting to a double representation might lose precision and allow the check below to succeed
+  // but the actual cast to nanoseconds fail. Using 1 DurationT worth of nanoseconds less than max
+  constexpr auto maximum_safe_cast_ns =
+    std::chrono::nanoseconds::max() - std::chrono::duration<DurationRepT, DurationT>(1);
+  // If period is greater than nanoseconds::max(), the duration_cast to nanoseconds will overflow
+  // a signed integer, which is undefined behavior. Checking whether any std::chrono::duration is
+  // greater than nanoseconds::max() is a difficult general problem. This is a more conservative
+  // version of Howard Hinnant's (the <chrono> guy>) response here:
+  // https://stackoverflow.com/a/44637334/2089061
+  // However, this doesn't solve the issue for all possible duration types of period.
```

Comment:
I mean, isn't that case detected in the method an a `std::invalid_argument` exception thrown?
If we always throw `std::invalid_argument` when the conversion is not possible we should comment that instead.
If there's UB under some conversions, we should clarify which cases are UB.

Reply:
Do you means we should show available range in comments ?  
e.g.  

If input time > 2562047788015 hours (nanoseconds::max) or input time < -2562047788015 hours (nanoseconds::min), conversion fails and exception std::invalid_argument is thrown.

Reply:
If input time > 2562047 hours (nanoseconds::max) or input time < -2562047 hours (nanoseconds::min), conversion fails and exception std::invalid_argument is thrown.

Reply:
> Do you means we should show available range in comments ?

No, that's already clear [here](https://github.com/ros2/rcpputils/blob/c6acdfaaa8db15f5a0adf0878ccae6a45dd83f77/include/rcpputils/time.hpp#L32-L33).

What I don't understand is the comment "However, this doesn't solve the issue for all possible duration types of period.". 
What duration types doesn't this solve the issue for? What is the "issue"?
If there's a duration type where the checks we're doing aren't going to guarantee that the conversion will either succeed or throw `std::invalid_argument`, we should be specifying for what duration types that can happen.

If the function is always going to either succeed or throw, and there's no posibility of UB for all possible duration types, then we should delete that comment because it's not really clear what it means.

Reply:
Thank you for the detailed explanation.    
Yes. This function return success or throw. There are  no other cases.    
I will update comments.